### PR TITLE
Authentication of Tokens

### PR DESCRIPTION
**Pull request by pastorsj**
#MINOR#

# CHANGELOG
* Added Passport authentication to protect routes and bolster login system
* Updated contract route to include a `fake` contract and a way to generate a token for a developer
* Customized denied route to include 5+ reasons why a user was denied access to the server
* Enabled the ability for authenticated users to submit tar balls with e2e tests to the server using their GitHub username and token
* Secured the backend from unauthorized people attempting to send malicious tarballs

### CODE REVIEW

File: app/controllers/auth/contract.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  checked: false,
+  confirmed: false,
+  disabled: true,
+  token: '',
+  actions: {
+    onInputHandler: function () {
+      this.set('checked', !this.get('checked'))
+      this.set('disabled', !this.get('checked'))
+    },
+    submitForm: function () {
```
Comment (pastorsj) [inline app/controllers/auth/contract.js]: jsdoc

File: app/controllers/auth/contract.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import Ember from 'ember'
+import generateToken from '../../utils/generateToken'
+
+export default Ember.Controller.extend({
+  checked: false,
+  confirmed: false,
+  disabled: true,
+  token: '',
+  actions: {
+    onInputHandler: function () {
```
Comment (pastorsj) [inline app/controllers/auth/contract.js]: jsdoc

File: app/routes/auth/contract.js [hunk 3]
```diff
@@ -1,4 +1,45 @@
...
+const ALREADY_HAS_ACCOUNT = 4
+const ERROR_ROUTE = 0
+
 export default Ember.Route.extend({
+  queryParams: {
+    username: {
+      refreshModel: true
+    }
+  },
+  model: function (params) {
```
Comment (pastorsj) [inline app/routes/auth/contract.js]: jsdoc

File: app/routes/auth/contract.js [hunk 4]
```diff
@@ -1,4 +1,45 @@
...
+          this.transitionTo('auth.denied', {queryParams: {reason: USER_DOES_NOT_EXIST}})
+        } else if (err.errors[0].status === '520') {
+          this.transitionTo('auth.denied', {queryParams: {reason: ALREADY_HAS_ACCOUNT}})
+        } else {
+          this.transitionTo('auth.denied', {queryParams: {reason: ERROR_ROUTE}})
+        }
+        Ember.Logger.debug(err)
+      })
+  },
+  setupController: function (controller, model) {
```
Comment (pastorsj) [inline app/routes/auth/contract.js]: jsdoc

File: app/routes/auth/denied.js [hunk 5]
```diff
@@ -1,4 +1,50 @@
...
 
+const NO_REASON = 6
+
 export default Ember.Route.extend({
+  queryParams: {
+    reason: {
+      refreshModel: true
+    }
+  },
+  model: function (params) {
```
Comment (pastorsj) [inline app/routes/auth/denied.js]: jsdoc

File: app/routes/auth/denied.js [hunk 6]
```diff
@@ -1,4 +1,50 @@
...
+    if (params.reason) {
+      return params.reason
+    }
+    return NO_REASON
+  },
+  setupController: function (controller, model) {
```
Comment (pastorsj) [inline app/routes/auth/denied.js]: jsdoc

File: app/routes/login.js [hunk 7]
```diff
@@ -1,9 +1,14 @@
...
 export default Ember.Route.extend({
-  beforeModel: function () {
-    if (window) {
-      window._isAuthenticated_ = false
+  queryParams: {
+    failure: {
+      refreshModel: true
+    }
+  },
+  model: function (params) {
```
Comment (pastorsj) [inline app/routes/login.js]: jsdoc

File: app/routes/auth/contract.js [hunk 8]
```diff
@@ -1,4 +1,45 @@
...
+        username: params.username,
+        token: '!'
+      })
+      .then((res) => {
+        return {
+          username: params.username
+        }
+      })
+      .catch((err) => {
+        if (err.errors[0].status === '510') {
```
Comment (psbanka) [inline app/routes/auth/contract.js]: These custom Error-codes are a Bad Idea. The situation you have on your hands is a `403`, which semantically is `unauthorized`. You should simply return a `403`. Extra details (e.g. 'user does not exist' or 'user already has an account') should be provided by the server and not by the client.

File: app/controllers/portal.js [hunk 9]
```diff
@@ -188,6 +190,30 @@ export default Ember.Controller.extend({
...
+    },
+    /**
+     * Logs the user out by calling the logout route. This route is being
+     * listened to by Passport and will logout the user and end the session. Then it will
+     * return successfully and we will transition to the index page.
+     */
+    logout: function () {
+      Ember.$.get('/logout', () => {
+        this.transitionToRoute('index')
+      })
```
Comment (psbanka) [inline app/controllers/portal.js]: After hitting logout and then navigating to an authorized URL, I think you found that the server was either (a) not invalidating the user's token when logging out, or (b) not checking the user's token properly when visiting a page that requires authentication. Please check this and add some tests to verify you don't have a regression.

Comment (pastorsj) [inline app/controllers/portal.js]: I fixed this issue in a recent commit. The session was not getting destroyed property.

File: app/controllers/auth/contract.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    submitForm: function () {
+      this.set('confirmed', true)
+      const username = this.get('username')
+      const token = generateToken(30)
+      this.set('username', username)
+      this.set('token', token)
+      return this.get('store').createRecord('developer', {
+        username,
+        token
+      })
```
Comment (pastorsj) [inline app/controllers/auth/contract.js]: Yes, working on it now

File: app/routes/auth/denied.js [hunk 11]
```diff
@@ -1,4 +1,50 @@
...
+    controller.set('reason', model)
+    switch (model) {
```
Comment (psbanka) [inline app/routes/auth/denied.js]: It seems like a weird design that the client needs to know this stuff. The server should send back some JSON to the client indicating that the attempt was successful or not and if not what the additional information is about the failure. Having to code this in both the server and the client is Too Tightly Coupled.

File: spec/api-spec.js [hunk 12]
```diff
@@ -253,34 +247,38 @@ describe('Delete Requests', () => {
...
-            })
-        }
+      .then((result) => {
+        expect(res.status).toHaveBeenCalledWith(200)
+        server.delete(reqDelete, res)
+          .then((result) => {
+            expect(res.status).toHaveBeenCalledWith(200)
+            server.get(reqGet, res)
+              .then((result) => {
+                console.log('1')
```
Comment (psbanka) [inline spec/api-spec.js]: left in gratuitious logging. Also other instances of this in the file.

File: src/app.js [hunk 13]
```diff
@@ -31,8 +38,8 @@ const github = new GitHubAPI({
   timeout: 5000
 })
 
-// const baseUrl = 'localhost:3000'
-const deploymentURL = 'wdio.bp.cyaninc.com'
+const baseUrl = 'localhost:3000'
+// const deploymentURL = 'wdio.bp.cyaninc.com'
```
Comment (psbanka) [inline src/app.js]: Why is this in here? Can it be removed?

File: src/app.js [hunk 14]
```diff
@@ -100,7 +143,7 @@ app.use('/screenshots', express.static(path.join(__dirname, '..', 'screenshots')
 app.get('/auth', function (req, res) {
   res.writeHead(303, {
     Location: OAuth2.getAuthorizeUrl({
-      'redirect_uri': 'http://' + deploymentURL + '/auth/callback',
+      'redirect_uri': 'http://' + baseUrl + '/auth/callback',
```
Comment (psbanka) [inline src/app.js]: why is this needed?

File: spec/api-spec.js [hunk 15]
```diff
@@ -123,38 +123,35 @@ describe('Get Requests', () => {
...
     server.get(req, res)
-      .then((err) => {
-        expect(err).not.toBe(null)
-        expect(res.status).toHaveBeenCalledWith(500)
+      .then((result) => {
+        done.fail(result)
+      })
+      .catch((errResult) => {
+        expect(errResult).not.toBe(null)
+        expect(res.status).toHaveBeenCalledWith(520)
```
Comment (psbanka) [inline spec/api-spec.js]: Again with the `520` error-code. This is a Bad Idea. No need to invent your own error-codes here. This is a simple `403` condition. Use that error-code instead.

File: app/routes/auth/contract.js [hunk 16]
```diff
@@ -16,7 +16,7 @@ export default Ember.Route.extend({
     return this.get('store').queryRecord('developer',
       {
         username: params.username,
-        token: '~'
+        token: '!'
```
Comment (psbanka) [inline app/routes/auth/contract.js]: for magic values like this, it would be helpful to designate a constant for them. In other words:
```
const HAS_NOT_ACCEPTED_CONTRACT = '!'
const TOKEN_REVOKED = '~'
...
{
  username: params.username,
  token: HAS_NOT_ACCEPTED_CONTRACT
}
```
Would be much more readable and maintainable.

File: handlers/developers-handler.js [hunk 17]
```diff
@@ -99,8 +99,17 @@ var DeveloperHandler = {
         })
       } else {
         this.client.get(req.query.username, (err, redisResp) => {
-          getResponse(err, res, req, redisResp)
-          resolve(err)
+          let response = {}
+          console.log('Get')
+          console.log(res)
+          console.log('Reponse gotten')
```
Comment (psbanka) [inline handlers/developers-handler.js]: Remove gratuitous logging

File: handlers/developers-handler.js [hunk 18]
```diff
@@ -166,15 +175,29 @@ function getKeysResponse (err, res, keys) {
...
+    if (res) {
+      DeveloperHandler.setErrorResponse(res, 'The username provided does not match any username. ' +
                                            'Please make sure that you are signed up as an authorized ' +
                                            'ciena developer on www.cienadevelopers.com', 510)
+    }
+    return {reject: 'Error, this username does not exist: ' + req.query.username}
   } else if (req.query.token === '' || req.query.token === redisResp) {
-    DeveloperHandler.setStandardResponse(res, req.query.username, redisResp)
+    console.log('Token' + req.query.token)
+    console.log(req.query.token)
```
Comment (psbanka) [inline handlers/developers-handler.js]: Gratuitous logging

File: src/process-upload.js [hunk 19]
```diff
@@ -86,11 +88,52 @@ const watchChild = function (child, seconds) {
...
+      }
+    })
+  }
+}
 
-  const child = childProcess.spawn('bash', [this.scriptPath, filename, entryPoint, seconds, testsFolder])
-  watchChild(child, seconds)
-  res.send(seconds.toString())
-  res.end()
+var getStatsofDirectoriesAndFolders = function (dir) {
```
Comment (psbanka) [inline src/process-upload.js]: jsdoc

**Comment by psbanka** [comment]
approved
